### PR TITLE
Config changes

### PR DESCRIPTION
**Pull request by ghost**
(Yes, Forge has a new config system, but Concrete's is much simpler in my opinion.)

* Improve documentation
* Allow private and/or inherited variables
* Add config GUI
    * Add `modID` parameter to `ConcreteConfig` constructor; used to handle `OnConfigChangedEvent`
    * Cache default config values
    * Add GUI-related variables to `ConfigValue`
    * Add `ConcreteGuiFactory`

### CODE REVIEW

**Review by exaskye — APPROVED** [review]
*bursts through wall*

LOOKS GOOD TO ME

**Comment by darkevilmac** [comment]
Seems like a lot of change for Config GUI support... Is there a way this could be done that's a bit smaller?

File: src/main/java/com/elytradev/concrete/config/ConcreteConfig.java [hunk 1]
```diff
@@ -63,83 +85,141 @@ protected ConcreteConfig(File configFile) {
...
+							break;
+						}
+					}
+				}
+
+				if (property != null) {
+					property.setShowInGui(showValueInGui);
+					property.setLanguageKey(valueLangKey);
+
+					if (valueRequiresMcRestart) {
```
Comment (darkevilmac) [inline src/main/java/com/elytradev/concrete/config/ConcreteConfig.java]: Flip the calls around and remove the extra if statements.

File: src/main/java/com/elytradev/concrete/config/ConcreteConfig.java [hunk 2]
```diff
@@ -63,83 +85,141 @@ protected ConcreteConfig(File configFile) {
 	 * Saves config values, and writes to disk.
 	 */
 	public void loadConfig() {
-		Class clazz = this.getClass();
+		try {
+			if (fieldToDefaultValue == null) {
```
Comment (darkevilmac) [inline src/main/java/com/elytradev/concrete/config/ConcreteConfig.java]: Might make sense to move this to another method, just to make it pretty. 'cacheDefaultValue' would work.

File: src/main/java/com/elytradev/concrete/config/ConcreteGuiFactory.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+import net.minecraftforge.fml.client.IModGuiFactory;
+import net.minecraftforge.fml.client.config.GuiConfig;
+import net.minecraftforge.fml.client.config.IConfigElement;
+
+public class ConcreteGuiFactory implements IModGuiFactory {
+	private final Configuration config;
+	private final String modID;
+	private final String title;
+
+	protected ConcreteGuiFactory(ConcreteConfig concreteConfig, String title) {
```
Comment (darkevilmac) [inline src/main/java/com/elytradev/concrete/config/ConcreteGuiFactory.java]: Remove this, add documentation to the class that notes it requires an empty constructor for Forge to instantiate it.

**Comment by ghost** [comment]
Please check if you still want the changes you requested. As of the latest commit, I don't think they are necessary.

**Review by darkevilmac — APPROVED** [review]

**Pull request merged**